design/ErrorHandling.md

```diff
@@ -164,147 +163,265 @@ WebGPU access has been disabled for the page.
+
+The WebGPU implementation may choose not to fire the `uncapturederror` event for a given error, for example if it has fired too many times, too many times in a row, or with too many errors of the same kind.
+This is similar to how console warnings would work, and work today for WebGL.
+(In badly-formed applications, this mechanism can prevent the events from having a significant performance impact on the system.)
+
+**Unlike** error scoping, the `uncapturederror` event can only fire on the main thread (Window) event loop.
+
+```webidl
+[
+    Constructor(DOMString type, GPUUncapturedErrorEventInit gpuUncapturedErrorEventInitDict),
+    Exposed=Window
+]
+interface GPUUncapturedErrorEvent : Event {
+    readonly attribute GPUError error;
 };
-```
 
-A concrete example: When creating a buffer, the following logic applies:
+dictionary GPUUncapturedErrorEventInit : EventInit {
+    required DOMString message;
+};
 
- - `createBuffer` returns a `GPUBuffer` object `buffer` immediately.
- - A `GPUObjectStatusQuery` can be obtained by calling `device.getObjectStatus(buffer)`.
-   At a later time, that query resolves to a `GPUObjectStatus` that is one of:
-    - Creation succeeded (`"valid"`).
-    - Creation encountered a recoverable error (`"out-of-memory"`).
-      (The application can then choose to retry a smaller allocation of a *new* `GPUBuffer`.)
-    - Creation encountered another type of error out of the control of the application (`"invalid"`).
+// TODO: is it possible to expose the EventTarget only on the main thread?
+partial interface GPUDevice : EventTarget {
+    [Exposed=Window]
+    attribute EventHandler onuncapturederror;
+};
+```
 
-Regardless of any recovery efforts the application makes, if creation fails,
-the resulting object is invalid (and subject to error propagation).
+#### Example
 
-Checking the status of a `GPUBuffer` or `GPUTexture` is **not** required.
-It is only necessary if an application wishes to recover from recoverable errors such as out of memory.
-(If it does, it is responsible for avoiding using the invalid object.)
+```js
+const device = await adapter.createDevice({});
+device.addEventListener('uncapturederror', (event) => {
+  appendToTelemetryReport(event.message);
+});
+```
 
 ## Open Questions and Considerations
 
- - WebGPU could guarantee that objects such as `GPUQueue` and `GPUFence` can never be errors.
-   If this is true, then the only synchronous API that needs special casing is buffer mapping, where `mapping` is always `null` for an invalid `GPUBuffer`.
+ - Is there a need for synchronous, programmatic capture of errors during development?
+   (E.g. an option to throw an exception on error instead of surfacing the error asynchronously.
+   Asynchronous error handling APIs are not enough to polyfill this.)
+   This would only be needed for printf-style debugging; a "break on WebGPU error" would be used for Dev Tools debugging.
```

Comment:
I think the most important issue this spec should solve is giving the developer the ability to determine the exact line of code where they've screwed up.  Throwing exceptions on errors is the only way to solve this, in my opinion. As you say, we can't polyfil this with asynchronous APIs. 

I would be satisfied with a dev-tools-only mode where exception-on-errors can be enabled, though it's tricky to test dev-tools-only things if you can't run them on from script.  If we allow developers the ability to turn on exceptions-on-errors from script, then we can test the API and give developers good debugging tools at the same time.  At that point, having a `pushErrorScope` and `popErrorScope` API becomes less important.

Reply:
Do you think there's a need for synchronous-Exception-on-error, rather than just break-debugger-on-error?

Reply:
(The asynchronous APIs *can* tell you exactly where you've screwed up, if you put an error scope around every single call e.g. in a debug layer. They just can't do so synchronously.)

Reply:
I like a dev-tools-only mode for that, and agree we should have one eventually.